transport/owr_transport_agent.c

```diff
@@ -1741,6 +1840,119 @@ static void on_component_state_changed(NiceAgent *nice_agent, guint stream_id,
+            g_warn_if_fail(sync_ok);
+
+            pad = gst_element_get_static_pad(pending_session_info->nice_src_rtp, "src");
+            gst_pad_remove_probe(pad, pending_session_info->nice_src_block_rtp);
+            gst_object_unref(pad);
+            pending_session_info->nice_src_block_rtp = 0;
+        } else if (component_id == NICE_COMPONENT_TYPE_RTCP && pending_session_info->nice_src_block_rtcp) {
+            GstPad *pad;
+            gboolean sync_ok, link_ok;
+
+            link_ok = gst_element_link(pending_session_info->nice_src_rtcp, pending_session_info->dtls_dec_rtcp);
+            g_warn_if_fail(link_ok);
+
+            gst_element_set_locked_state(pending_session_info->dtls_dec_rtcp, FALSE);
+            sync_ok = gst_element_sync_state_with_parent(pending_session_info->dtls_dec_rtcp);
+            g_warn_if_fail(sync_ok);
+            gst_element_set_locked_state(pending_session_info->dtls_enc_rtcp, FALSE);
+            sync_ok = gst_element_sync_state_with_parent(pending_session_info->dtls_enc_rtcp);
+            g_warn_if_fail(sync_ok);
+
+            pad = gst_element_get_static_pad(pending_session_info->nice_src_rtcp, "src");
+            gst_pad_remove_probe(pad, pending_session_info->nice_src_block_rtcp);
+            gst_object_unref(pad);
+            pending_session_info->nice_src_block_rtcp = 0;
+        }
+    }
+    g_mutex_unlock(&transport_agent->priv->sessions_lock);
+}
+
+static gboolean maybe_handle_new_send_source_with_payload_from_main_thread(GHashTable *args)
+{
+    OwrTransportAgent *transport_agent;
+    OwrMediaSession *session;
+
+    session = OWR_MEDIA_SESSION(g_hash_table_lookup(args, "session"));
+    g_return_val_if_fail(OWR_IS_MEDIA_SESSION(session), FALSE);
+    transport_agent = OWR_TRANSPORT_AGENT(g_hash_table_lookup(args, "transport_agent"));
+    g_return_val_if_fail(OWR_IS_TRANSPORT_AGENT(transport_agent), FALSE);
+
+    maybe_handle_new_send_source_with_payload(transport_agent, session);
+
+    g_hash_table_destroy(args);
+    g_object_unref(session);
+    g_object_unref(transport_agent);
+
+    return FALSE;
+}
+
+static void
+on_dtls_enc_key_set(GstElement *dtls_srtp_enc, AgentAndSessionIdPair *data)
+{
+    OwrTransportAgent *transport_agent = data->transport_agent;
+    OwrSession *session;
+    guint stream_id = data->session_id;
+    PendingSessionInfo *pending_session_info;
+
+    session = get_session(transport_agent, stream_id);
+    g_return_if_fail(session);
+
+    /* Once we have the key, the DTLS handshake is done and we can start
+     * sending data here
+     */
+    g_mutex_lock(&transport_agent->priv->sessions_lock);
+    pending_session_info = g_hash_table_lookup(transport_agent->priv->pending_sessions, GUINT_TO_POINTER(stream_id));
+    /* FIXME: What to do about RTCP? It's not guaranteed to ever be enabled if
+     * RTCP muxing is used but the usage wasn't known beforehand
```

Comment:
I *think* what I meant here is that we only wait for RTP, not RTCP (or both). So in theory we might have finished setup for RTP, then put everything together, and RTCP is still not ready yet and RTCP packets get lost.
This is only a problem if no RTCP muxing is used.